[Unit-7]Basic chapter of video and video processing (#165)
johko/computer-vision-course: chapters/en/Unit 7 - Video and Video Processing/video-processing-basics.mdx

Temporal relations are the main factors for developing a strong understanding of content in a video, thus we require a separate algorithm for videos. One of the main differences between image and video processing is the inclusion of an additional axis, time to the input. 
There are two main approaches for extracting tokens from video or embedding a video clip:

### Uniform Frame Sampling:
It is a straightforward method of tokenizing the input video in which we uniformly sample 'nt' frames from the input video clip, embed each 2D frame independently using the same method as used in image processing and concatenate all these tokens together.
If 'nh.nw' non-overlapping image patches are extracted from each frame, then a total of 'nt.nh.nw' tokens will be forwarded through the transformer encoder. Uniform frame sampling is a tokenization scheme in which we sample frames from the video clip and perform simple ViT tokenization.

### Tubelet Embedding:

> This method is an extension of the vision transformer's image embedding to 3D and corresponds to a 3D convolution. It is an alternative method in which non-overlapping, spatio-temporal “tubes” from input volume are extracted and linearly projected.

Reviewer: ```suggestion
This method is an extension of the vision transformer's image embedding to 3D and corresponds to a 3D convolution. It is an alternative method in which non-overlapping, spatiotemporal “tubes” from input volume are extracted and linearly projected.

```

> First, we extract tubes from the video, these tubes contain patches of the frame and the temporal information as well. The tubes are then flattened to build video tokens. Intuitively, this method fuses spatio-temporal information during tokenization, in contrast to “Uniform frame sampling” where temporal information from different frames is fused by the transformer.

Reviewer: ```suggestion
First, we extract tubes from the video, these tubes contain patches of the frame and the temporal information as well. The tubes are then flattened to build video tokens. Intuitively, this method fuses spatiotemporal information during tokenization, in contrast to “Uniform frame sampling” where temporal information from different frames is fused by the transformer.

```

# Video Processing Basics

> With the rise and advancements of transformers, vision transformers have become essential tools for various computer vision tasks. Vision transformers are great at computer vision tasks involving both images and videos.

Reviewer 2: ```suggestion
With the rise of transformers, vision transformers have become essential tools for various computer vision tasks. Vision transformers are great at computer vision tasks involving both images and videos.

```

Suppose we have a 224x224 image, split the image into 16x16 patches where each patch consists of 14x14 pixels. This patch-based processing not only reduces computation but it also allows the model to capture local features in the image effectively.
Each patch is then fed through a series of self-attention layers and feed-forward neural networks to extract semantic information. Thanks to this hierarchical processing technique, vision transformers are able to capture both high-level and low-level features in the image.

Since vision transformers process patches individually and transformers by default don't have any mechanism to track position of inputs, the spatial context of the image can be lost.
To address this, vision transformers often include positional encodings that capture the relative position of each patch within the image. By incorporating positional information, the model can better understand the spatial relationships between different patches and enhance its ability to recognize objects and patterns.

> 

Reviewer 2: ```suggestion
```

*__Note__: CNNs are designed to learn spatial features, while vision transformers are designed to learn both spatial and contextual features.*

## Key Differences Between Image and Video Processing

> Videos are essentially a sequence of frames, and processing them requires techniques to capture and incorporate motion information. In image processing, the transformer ignores the temporal (time) relations between frames i.e. it only focuses on spatial (space) information of a frame.

Reviewer 2: ```suggestion
Videos are essentially a sequence of frames, and processing them requires techniques to capture and incorporate motion information. In image processing, the transformer ignores the temporal (time) relations between frames, i.e., it only focuses on a frame's spatial (space) information.

```
